src/checker.ts

```diff
@@ -442,12 +443,43 @@ export function makeFunctionlessChecker(
     return false;
   }
 
+  /**
+   * Checks if the {@link id} references a variable
+   * @param id
+   * @returns
+   */
+  function isIdentifierReference(id: ts.Identifier) {
+    if (ts.isBindingElement(id.parent)) {
+      // { a: b = c }
+      //   ^T ^T  ^F
+      return id.parent.initializer === id;
+    } else if (ts.isParameter(id.parent)) {
+      // function foo(a = b)
+      //              ^F  ^T
+      return id.parent.initializer === id;
+    } else if (ts.isVariableDeclaration(id.parent)) {
+      // const a = b;
+      //       ^F  ^T
+      return id.parent.initializer === id;
+    } else if (ts.isPropertyAccessExpression(id.parent)) {
+      // event.bus
+      // ^T    ^F
+      return id.parent.expression === id;
+    }
+
+    return true;
+  }
+
```

Comment:
Why not use the flags? Those seems to be correct? 

* https://github.com/functionless/functionless/blob/main/src/checker.ts#L774

or

```ts
                symbol?.valueDeclaration &&
                // eslint-disable-next-line no-bitwise
                (symbol.flags & ts.SymbolFlags.BlockScopedVariable ||
                  // eslint-disable-next-line no-bitwise
                  symbol.flags & ts.SymbolFlags.FunctionScopedVariable)
```

If not...

This seems like the wrong approach, to check how it is being used, missing other cases like being used in a binary expression, switch statement, conditional statement, return statement, and more.

Reply:
Those cases don't matter - this function all analyzes them correctly. Not sure why this is the wrong approach?

I was unaware of the flags. Let me check on that

Reply:
The problem is we are treating all identifiers as references and they are not. Why is this logic incorrect? The parent defines the interpretation of the identifier

Reply:
If we keep this logic, should document which cases it is for and maybe update the name of the method to reflect that.

`isIdentifierUsedAsAVariableInitializer`?

also this seems backwards?

```ts
// { a: b = c }
//   ^T ^T  ^F
```

and what about ?

// const a = x ? b : c

Reply:
```ts
const a = x ? b : c
```

this would return false for a and true for the rest

Reply:
> also this seems backwards?

Typo I think. My bad

Reply:
Ohh, I see, you are encoding all of the cases where ident is used as a name instead of a variable. (I am an idiot and missed the `return true`). I think the flags are going to be the most accurate, they seemed to work for me without any new logic.

Reply:
I don't think flags work as `c` in the below code is a BlockScopedVariable
```ts
const d = "";
const {b: c = d} = {b: ""};
```